dpcpp/solver/batch_bicgstab_launch.instantiate.dp.cpp

```diff
@@ -0,0 +1,112 @@
+#include "core/solver/batch_dispatch.hpp"
+#include "dpcpp/base/batch_multi_vector_kernels.hpp"
+#include "dpcpp/base/batch_struct.hpp"
+#include "dpcpp/base/config.hpp"
+#include "dpcpp/base/dim3.dp.hpp"
+#include "dpcpp/base/dpct.hpp"
+#include "dpcpp/base/helper.hpp"
+#include "dpcpp/components/cooperative_groups.dp.hpp"
+#include "dpcpp/components/intrinsics.dp.hpp"
+#include "dpcpp/components/reduction.dp.hpp"
+#include "dpcpp/components/thread_ids.dp.hpp"
+#include "dpcpp/solver/batch_bicgstab_kernels.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace dpcpp {
+namespace batch_bicgstab {
+
+
+template <typename ValueType, typename StopType, const int subgroup_size,
+          const int n_shared_total, typename PrecType, typename LogType,
+          typename BatchMatrixType>
+void launch_apply_kernel(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const gko::kernels::batch_bicgstab::storage_config& sconf,
+    const settings<remove_complex<ValueType>>& settings, LogType& logger,
+    PrecType& prec, const BatchMatrixType& mat,
+    const ValueType* const __restrict__ b_values,
+    ValueType* const __restrict__ x_values,
+    ValueType* const __restrict__ workspace, const int& group_size,
+    const int& shared_size)
+{
+    auto num_rows = mat.num_rows;
+
+    const dim3 block(group_size);
+    const dim3 grid(mat.num_batch_items);
+
+    auto max_iters = settings.max_iterations;
+    auto res_tol = settings.residual_tol;
+
+    exec->get_queue()->submit([&](sycl::handler& cgh) {
+        sycl::accessor<ValueType, 1, sycl::access_mode::read_write,
+                       sycl::access::target::local>
+            slm_values(sycl::range<1>(shared_size), cgh);
+
+        cgh.parallel_for(
+            sycl_nd_range(grid, block),
+            [=](sycl::nd_item<3> item_ct1) [[intel::reqd_sub_group_size(
+                subgroup_size)]] [[intel::kernel_args_restrict]] {
+                auto batch_id = item_ct1.get_group_linear_id();
+                const auto mat_global_entry =
+                    gko::batch::matrix::extract_batch_item(mat, batch_id);
+                const ValueType* const b_global_entry =
+                    gko::batch::multi_vector::batch_item_ptr(
+                        b_values, 1, num_rows, batch_id);
+                ValueType* const x_global_entry =
+                    gko::batch::multi_vector::batch_item_ptr(
+                        x_values, 1, num_rows, batch_id);
+                batch_single_kernels::apply_kernel<StopType, n_shared_total>(
+                    sconf, max_iters, res_tol, logger, prec, mat_global_entry,
+                    b_global_entry, x_global_entry, num_rows,
+                    mat.get_single_item_num_nnz(),
+                    static_cast<ValueType*>(slm_values.get_pointer()), item_ct1,
+                    workspace);
+            });
+    });
+}
+
+
+// begin
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_BATCH_BICGSTAB_LAUNCH_0);
```

Comment:
It will not work with `GINKGO_DPCPP_SINGLE_MODE=ON`.
we use the instantiation to provide the specialization with unsupported exception on double precision.
with `GKO_BATCH_INSTANTION`, it will be wrong. Only the last one has the specialization, but the others will be instantiated.
`template macro {GKO_UNSUPPORTED;}` ->
```
template first_...;
template second_...;
...
template last  {GKO_UNSUPPORTED;}
```

Reply:
You're right. Thanks for bringing this up. I changed the order of macro application, so now it should be fixed.